feat(synthetics): Synthetics L2 Support (#8824)
aws/aws-cdk: packages/@aws-cdk/aws-synthetics/README.md

---
<!--END STABILITY BANNER-->

Amazon CloudWatch Synthetics allow you to monitor your application by generating **synthetic** traffic. The traffic is produced by a **canary**: a configurable script that runs on a schedule. You configure the canary script to follow the same routes and perform the same actions as a user, which allows you to continually verify your user experience even when you don't have any traffic on your applications.

## Canary

To illustrate how to use a canary, assume your application defines the following endpoint:

```bash
% curl "https://api.example.com/user/books/topbook/"
The Hitchhikers Guide to the Galaxy

```

The below code defines a canary that will hit the `books/topbook` endpoint every 5 minutes: 

```ts
import * as synthetics from '@aws-cdk/aws-synthetics';

const canary = new synthetics.Canary(this, 'MyCanary', {
  schedule: synthetics.Schedule.rate(Duration.minutes(5)),
  test: Test.custom({
    code: Code.fromAsset(path.join(__dirname, 'canary'))),
    handler: 'index.handler',
  }),
});
```

> The following is an example of a `canary/index.js` which exports the `handler` function:

Reviewer: ```suggestion
The following is an example of an `index.js` file which exports the `handler` function, note that the function **must** be called `handler`:
```

```
const https = require('https');

> var synthetics = require('Synthetics');

Reviewer: This is imported but never used - what would be a usage of this?

Reviewer 2: Wouldn't it be better to show an example that really uses the `Synthetics` library? Or at least mention somewhere in the README that it's available (and includes Puppeteer and Chromium). Its one of the main features of Synthetics Canaries. The *Heartbeat monitoring* blueprint of the console experience demonstrates the "power" of this very well.

Author: I've removed the current example and added in the heartbeat monitoring example. Should I be explicitly mentioning the `Synthetics` library elsewhere in the readme? If this is a big selling point for Canaries, then perhaps the README should highlight it more than just in this example.

const log = require('SyntheticsLogger');

exports.handler = async function () {
  const requestOptions = {"hostname":"api.example.com","method":"","path":"/user/books/topbook/","port":443}
  let req = https.request(requestOptions);

>   req.on('response', (res) => {

Reviewer: Is there any special handling needed if the request was unsuccessful? Both in terms of connection errors and bad responses (>= 400)

>     log.info()

Reviewer: This seems funky, add a log message?

  });
}
```

The canary will automatically produce a CloudWatch Dashboard:

![UI Screenshot](images/ui-screenshot.png)

### Configuring the Canary Script 

> To configure the script the canary executes, use the `test` property. The `test` property exposes `code` and `handler` properties -- both are required by Synthetics to create a lambda function on your behalf. 

Reviewer: The `test` property doesn't expose the `code` and `handler` properties, `Test.custom` does.

```suggestion
To configure the script the canary executes, use the `test` property. Currently, you can only use `Test.custom` as the value, which lets you configure the `code` and the `handler` properties that create a lambda function on your behalf to run your custom code. 
```

Reviewer 3: I don't like the "you can only use `Test.custom` as the value" let me think here

Reviewer 3: To configure the script the canary executes, use the `test` property. The `test` property accepts a `Test` instance that can be initialized by the `Test` class static methods, currently the only implemented method is `Test.custom()`, which allows you to bring your on code. In the future other methods will be added. `Test.custom` accepts `code` and `handler` properties -- both are required by Synthetics to create a lambda function on your behalf.

The `synthetics.Code` class exposes static methods to bundle your code artifacts: 

  - `code.fromInline(code)` - specify an inline script.
  - `code.fromAsset(path)` - specify a .zip file or a directory in the local filesystem which will be zipped and uploaded to S3 on deployment. See note for directory structure.
  - `code.fromBucket(bucket, key[, objectVersion])` - specify an S3 object that contains the .zip file of your runtime code. See note for directory structure.

> > **Note:** For `code.fromAsset()` and `code.fromBucket()`, the canary resource requires the following folder structure: `nodejs/node_modules/<handlerFile>`. See Synthetics [docs](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_WritingCanary.html).

Reviewer: I would move this up just after the example of `index.js`. And use standard directory structure ASCII:

```text
canary/
├── nodejs/
    ├── node_modules/
        ├── index.js
```

Reviewer 3: lovely

> To supply the code from your local filesystem:

Reviewer: You actually already showed this example in the beginning, which is good. This doesn't really any additional benefit. Maybe show the `code.fromInline` here?

Reviewer 3: Yeah I agree, we can have a code  block showing all of the options in brevity

## Canary

To illustrate how to use a canary, assume your application defines the following endpoint:

```bash
curl https://api.awsomesite.com/user/books/topbook/
The Hitchhikers Guide to the Galaxy

```

The bellow code defines a canary that will hit the `books/topbook` endpoint every 5 minutes:

```ts
const canary = new Canary(this, 'MyCanary', {
  name: 'mycanary'
  schedule: Schedule.Rate(Duration.minutes(1)),
  handler: 'index.handler',
  code: synth.Code.fromInline(
    `const https = require('https');
     var synthetics = require('Synthetics');
     const log = require('SyntheticsLogger');
  
     exports.handler = async function () {
      const requestOptions = {"hostname":"api.awsomesite.com","method":"","path":"/user/books/topbook/","port":443}
      let req = https.request(requestOptions);
      req.on('response', (res) => {
        log.info()
      });
     }`
    )
});

> ```

Author: Sounds good, I like it.

> ```

Author: Sounds good, I like it.

The canary will automatically produce a CloudWatch Dashboard:

![UI Screenshot](images/ui-screenshot-1.png)

### Alarms

You can configure a CloudWatch Alarm on canary metrics. Metrics are emitted by CloudWatch automatically and can be accessed by the following APIs:
- `canary.metricSuccessPercent()`
- `canary.metricDuration()`
- `canary.metricFailed()`

To configure an alarm use the `addAlarm()` method:
```ts

canary.addAlarm('CanaryAlarm', {
  metric: canary.metricSuccessPercent(),
  evaluationPeriods: 2,
  threshold: 90,
  comparisonOperator: cloudwatch.ComparisonOperator.LESS_THAN_THRESHOLD,
})

> ```

Author: I've been thinking about this and I think you're right. My proposed `addAlarm` method simply creates an `Alarm` with the canary as a scope, and it can be done in a similar fashion without this method. 

The only 'pro' I can think of is that this nudges the customer in the direction of creating an alarm on a canary as that is the preferred use case. I can imagine that a common pattern will be to create a canary and then an alarm on it, so I thought it would make sense to acknowledge this pattern. But I agree, it is not needed.

 I'll leave it on the readme for now but I certainly have no qualms about removing it.

---
<!--END STABILITY BANNER-->

Amazon CloudWatch Synthetics allow you to monitor your application by generating **synthetic** traffic. The traffic is produced by a **canary**: a configurable script that runs on a schedule. You configure the canary script to follow the same routes and perform the same actions as a user, which allows you to continually verify your user experience even when you don't have any traffic on your applications.

## Canary

To illustrate how to use a canary, assume your application defines the following endpoint:

```bash
% curl "https://api.example.com/user/books/topbook/"
The Hitchhikers Guide to the Galaxy

```

The below code defines a canary that will hit the `books/topbook` endpoint every 5 minutes: 

```ts
import * as synthetics from '@aws-cdk/aws-synthetics';

const canary = new synthetics.Canary(this, 'MyCanary', {
  schedule: synthetics.Schedule.rate(Duration.minutes(5)),
  test: Test.custom({
    code: Code.fromAsset(path.join(__dirname, 'canary'))),
    handler: 'index.handler',
  }),
});
```

> The following is an example of an `index.js` file which exports the `handler` function, note that the function **must** be called `handler`:

Reviewer 3: ```suggestion
The following is an example of an `index.js` file which exports the `handler` function:
```

```js
var synthetics = require('Synthetics');
const log = require('SyntheticsLogger');

const pageLoadBlueprint = async function () {

    // INSERT URL here
    const URL = "https://api.example.com/user/books/topbook/";

    let page = await synthetics.getPage();
    const response = await page.goto(URL, {waitUntil: 'domcontentloaded', timeout: 30000});
    //Wait for page to render.
    //Increase or decrease wait time based on endpoint being monitored.
    await page.waitFor(15000);

>     await synthetics.takeScreenshot('loaded', 'loaded');

Reviewer 3: ```suggestion
    await synthetics.takeScreenshot('loaded', 'loaded'); // This will take a snapshot that will be included in test output artifacts
```

    let pageTitle = await page.title();
    log.info('Page title: ' + pageTitle);
    if (response.status() !== 200) {
        throw "Failed to load page!";
    }
};

exports.handler = async () => {
    return await pageLoadBlueprint();
};

> ```

Reviewer 3: here you can add a note section
```
> the function must be named `handler` blah blah
```

The canary will automatically produce a CloudWatch Dashboard:

![UI Screenshot](images/ui-screenshot.png)

> 

Reviewer 3: Following up on the comments above, we can add a line here that elborate on the Synthetics capabilities:

```
The Canary code will be executed in a lambda function created by Synthetics on your behalf. The Lambda function includes a custom [runtime](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_Library.html) provided by Synthetics. The provided runtime includes a variety of handy tools such as [Puppeteer](https://www.npmjs.com/package/puppeteer-core)  and Chromium. To learn more about Synthetics capabilities, check out the [docs](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries.html)
```

> The APIs of higher level constructs in this module are experimental and under active development. They are subject to non-backward compatible changes or removal in any future version. These are not subject to the [Semantic Versioning](https://semver.org/) model and breaking changes will be announced in the release notes. This means that while you may use them, you may need to update your source code when upgrading to a newer version of this package.

---
<!--END STABILITY BANNER-->

Amazon CloudWatch Synthetics allow you to monitor your application by generating **synthetic** traffic. The traffic is produced by a **canary**: a configurable script that runs on a schedule. You configure the canary script to follow the same routes and perform the same actions as a user, which allows you to continually verify your user experience even when you don't have any traffic on your applications.

## Canary

To illustrate how to use a canary, assume your application defines the following endpoint:

```bash
% curl "https://api.example.com/user/books/topbook/"
The Hitchhikers Guide to the Galaxy

```

> The below code defines a canary that will hit the `books/topbook` endpoint every 5 minutes: 

Reviewer 2: The example has been adapted to show how to take a screenshot. The use case is not an API anymore.

Author: I wanted to show off the power of the synthetics library, as you mentioned in a previous comment. The heartbeat blueprint works for APIs as well, and takes a screenshot of the rendered HTML (in this case, would capture a screenshot of a white page with `The Hitchhikers Guide to the Galaxy`). I tested this on my own API endpoint to make sure. 

While the screenshot is likely more powerful for websites, it can still be a valuable addition to APIs.

    handler: 'index.handler',
  }),
});
```

The following is an example of an `index.js` file which exports the `handler` function:

```js
var synthetics = require('Synthetics');
const log = require('SyntheticsLogger');

const pageLoadBlueprint = async function () {

    // INSERT URL here
    const URL = "https://api.example.com/user/books/topbook/";

    let page = await synthetics.getPage();
    const response = await page.goto(URL, {waitUntil: 'domcontentloaded', timeout: 30000});
    //Wait for page to render.
    //Increase or decrease wait time based on endpoint being monitored.
    await page.waitFor(15000);
    // This will take a screenshot that will be included in test output artifacts
    await synthetics.takeScreenshot('loaded', 'loaded');
    let pageTitle = await page.title();
    log.info('Page title: ' + pageTitle);
    if (response.status() !== 200) {
        throw "Failed to load page!";
    }
};

exports.handler = async () => {
    return await pageLoadBlueprint();
};
```

> **Note:** The function **must** be called `handler`.

The canary will automatically produce a CloudWatch Dashboard:

![UI Screenshot](images/ui-screenshot.png)

> The Canary code will be executed in a lambda function created by Synthetics on your behalf. The Lambda function includes a custom [runtime](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_Library.html) provided by Synthetics. The provided runtime includes a variety of handy tools such as [Puppeteer](https://www.npmjs.com/package/puppeteer-core)  and Chromium. To learn more about Synthetics capabilities, check out the [docs](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries.html)

Reviewer 3: ```suggestion
The Canary code will be executed in a lambda function created by Synthetics on your behalf. The Lambda function includes a custom [runtime](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_Library.html) provided by Synthetics. The provided runtime includes a variety of handy tools such as [Puppeteer](https://www.npmjs.com/package/puppeteer-core)  and Chromium. To learn more about Synthetics capabilities, check out the [docs](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries.html).
```

### Configuring the Canary Script 

To configure the script the canary executes, use the `test` property. The `test` property accepts a `Test` instance that can be initialized by the `Test` class static methods. Currently, the only implemented method is `Test.custom()`, which allows you to bring your own code. In the future, other methods will be added. `Test.custom()` accepts `code` and `handler` properties -- both are required by Synthetics to create a lambda function on your behalf.

> > **Note:** For `code.fromAsset()` and `code.fromBucket()`, the canary resource requires the following folder structure:

Reviewer 3: This note should be under the code sample

>```
>canary/
>├── nodejs/
>   ├── node_modules/
>        ├── <filename>.js
>```
> See Synthetics [docs](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_WritingCanary.html).

The `synthetics.Code` class exposes static methods to bundle your code artifacts: 

  - `code.fromInline(code)` - specify an inline script.
  - `code.fromAsset(path)` - specify a .zip file or a directory in the local filesystem which will be zipped and uploaded to S3 on deployment. See the above Note for directory structure.
  - `code.fromBucket(bucket, key[, objectVersion])` - specify an S3 object that contains the .zip file of your runtime code. See the above Note for directory structure.

> 

Reviewer 3: Add a header here so it will be more structured:

```
Using the `Code` class static initializers:   
```

```ts
// To supply the code inline:
const canary = new Canary(this, 'MyCanary', {
  test: Test.custom({
    code: Code.fromInline('/* Synthetics handler code */'),
    handler: 'index.handler', // must be 'index.handler'
  }),
});

// To supply the code from your local filesystem:
const canary = new Canary(this, 'MyCanary', {
  test: Test.custom({
    code: Code.fromAsset(path.join(__dirname, 'canary')),
    handler: 'index.handler', // must end with '.handler'
  }),
});

// To supply the code from a S3 bucket:
const canary = new Canary(this, 'MyCanary', {
  test: Test.custom({
    code: Code.fromBucket(bucket, 'canary.zip'),
    handler: 'index.handler', // must end with '.handler'
  }),
}); 
```

> 

Reviewer 3: ```suggestion
> **Note:** For `code.fromAsset()` and `code.fromBucket()`, the canary resource requires the following folder structure
```

### Alarms

> You can configure a CloudWatch Alarm on canary metrics. Metrics are emitted by CloudWatch automatically and can be accessed by the following APIs:

Reviewer 3: ```suggestion
You can configure a CloudWatch Alarm on the canary metrics. Metrics are emitted by CloudWatch automatically and can be accessed by the following APIs:
```

Author: I changed it to "You can configure a CloudWatch Alarm on a canary metric"

## Amazon CloudWatch Synthetics Construct Library

<!--BEGIN STABILITY BANNER-->
---

![cfn-resources: Stable](https://img.shields.io/badge/cfn--resources-stable-success.svg?style=for-the-badge)

> All classes with the `Cfn` prefix in this module ([CFN Resources](https://docs.aws.amazon.com/cdk/latest/guide/constructs.html#constructs_lib)) are always stable and safe to use.

![cdk-constructs: Experimental](https://img.shields.io/badge/cdk--constructs-experimental-important.svg?style=for-the-badge)

> The APIs of higher level constructs in this module are experimental and under active development. They are subject to non-backward compatible changes or removal in any future version. These are not subject to the [Semantic Versioning](https://semver.org/) model and breaking changes will be announced in the release notes. This means that while you may use them, you may need to update your source code when upgrading to a newer version of this package.

---
<!--END STABILITY BANNER-->

> Amazon CloudWatch Synthetics allow you to monitor your application by generating **synthetic** traffic. The traffic is produced by a **canary**. A canary is a configurable script that runs on a schedule. You configure the canary script to follow the same routes and perform the same actions as a customer, which allows you to continually verify your customer experience even when you don't have any customer traffic on your applications.

Reviewer 4: You could join the sentences here. Also I'd use different wording sporadically:

 ```suggestion
Amazon CloudWatch Synthetics allow you to monitor your application by generating **synthetic** traffic. The traffic is produced by a **canary**: a configurable script that runs on a schedule. You configure the canary script to follow the same routes and perform the same actions as a user, which allows you to continually verify your user experience even when you don't have any traffic on your applications.
```

## Canary

To illustrate how to use a canary, assume your application defines the following endpoint:

```bash

> curl https://api.awsomesite.com/user/books/topbook/

Reviewer 5: `~` Nit - Prefer example.com as a domain name in examples, as it's a domain specifically set up for examples in documentation. Someone may register "awsomesite.com" at some point and host material we wouldn't want to point users to from our docs. Also, might want to show that the first line is a bash command with a prompt of some kind.

```suggestion
% curl "https://api.example.com/user/books/topbook/"
```

> curl https://api.awsomesite.com/user/books/topbook/

Reviewer 4: Be **extremely** careful with those - the domain here is not owned by AWS... It could be highjacked by a malicious party who'd be able to easily exploit vulnerabilities of `curl` if people attempt to run this command.

Allows use domains that you own/control in your examples, even if the sites don't exist.

Author: Changed to `example.com` per Nick's suggestion.

The Hitchhikers Guide to the Galaxy

```

The below code defines a canary that will hit the `books/topbook` endpoint every 5 minutes:

> > **Not Implemented:** the `test` property will be a part of Milestone 2.

Reviewer 5: You have these notes throughout; I think it would be better to have the README only reflect the "live" code, and stash the full examples in an RFC, design doc, etc. As-is, it's really hard to read through the README and understand what functionality is actually available.

Author: Yep, my original idea was more to provide context for **reviewers** that might be jumping into my code cold turkey (and neglected to mention that this PR is certainly a **draft**). I wanted to provide the full vision for the construct rather than just what was implemented -- but as you and @RomainMuller have pointed out, this was more confusing than anything. I will update the README to reflect the "live" code and resolve all of your issues on the poor documentation.

```ts
const canary = new Canary(this, 'MyCanary', {

>   name: 'mycanary'

Reviewer 4: Is this required? If not, omit it. If so, this smells bad.

Author: Not required anymore.

  frequency: Duration.minutes(5),

>   test: synth.Test.custom({

Reviewer 4: Unsure about the `synth` name here. People could mistake it for the `cdk synth` meaning? It's also not introduced, so it could be anything.

Author: I hadn't thought of that. I'll introduce the import as `import * as synthetics from '@aws-cdk/aws-synthetics';` and use `synthetics` instead of `synth` to avoid confusion.

>     code: synth.Code.fromInline(`const https = require('https');

Reviewer 4: Is the `synth.Code` the same than `@aws-cdk/aws-lambda.Code`? If they differ - is there a way they could be unified?

Reviewer 3: You mean moving `Code` to core?
The biggest difference is that in the synthetics module, if the code is stored in s3,  the synthetics `Code` property requires that the zip file will have a specific structure, from their [docs](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_WritingCanary.html):

>If you have more than a single .js file or you have a dependency that your script depends on, you can bundle them all into a single ZIP file that contains the folder structure nodejs/node_modules/myCanaryFilename.js file and other folders and files.

(Note that the "If you have more than" is misleading as this is always required)

If the zip file does not conform to the describe structure the deployment will succeed but **the canary execution will fail** , possibly triggering an alarm. To spare this from users the synthetics `Code` class should include the logic that zip the local directory (asset).

      var synthetics = require('Synthetics');
      const log = require('SyntheticsLogger');
  
      exports.handler = async function () {

>         const requestOptions = {"hostname":"api.awsomesite.com","method":"","path":"/user/books/topbook/","port":443}

Reviewer 5: ```suggestion
        const requestOptions = {"hostname":"api.example.com","method":"","path":"/user/books/topbook/","port":443}
```

        let req = https.request(requestOptions);
        req.on('response', (res) => {
          log.info()
        });
      }`),

>     handler: 'index.handler',

Reviewer 4: Are synthetics always using Node as a runtime?

Author: Yup :)

  }),
});
```

The canary will automatically produce a CloudWatch Dashboard:

![UI Screenshot](images/ui-screenshot.png)

### Canary Test Property

> The `test` property represents the test that the canary executes. It accepts a type `Test`, which has as parameters the properties required to generate the test script.

Reviewer 4: > It accepts a type `Test`, which has as parameters the properties required to generate the test script

You're explaining how it is implemented and re-hashing the type signature of the constructor. I don't think that's what people want to read here.

> **Not Implemented:** `Test.heartBeat()` is a part of Milestone 3

Testing URL heartbeat:

```ts
const canary = new Canary(this, 'MyCanary', {
  test: Test.heartBeat('https://myawsomesite.com'),
  // ...
});
```

> **Not Implemented:** `Test.apiEndpoint()` is a part of Milestone 4

Testing a specific API:

```ts
const canary = new Canary(this, 'MyCanary', {
  test: Test.apiEndpoint('https://myawsomesite.com/endpoint', {
    method: 'GET'|'POST',
    headers: Map<headerName, headerValue>,

>     data: string,

Reviewer 4: What is this even? It's not valid TypeScript (type signatures used as an object literal??)

Author: Am removing this for now but when I implement this API I will add in a valid TS example :).

  }),
  // ...
});
```

> **Not Implemented:** `Test.brokenLink()` is a part of Milestone 4

Testing broken links:

```ts
const canary = new Canary(this, 'MyCanary', {
  test: Test.brokenLink('https://myawsomesite.com', {
    maxLinks: number,
  }),
  // ...
});
```

> **Not Implemented:** `Test.custom` is a part of Milestone 2

Users will be able to supply their own code by using `Test.custom()`, which exposes the `code` property similarly to the lambda construct.

```ts
const canary = new Canary(this, 'MyCanary', {
  test: Test.custom({

>     code: Code.fromAsset()|Code.fromBucket()|Code.fromInline(),

Reviewer 4: Still not valid TypeScript.

    handler: 'index.handler' // required by the resource 
  }),
  // ...
});
```

### Alarms

> **Not Implemented:** metric APIs will be a part of Milestone 3

You can configure a CloudWatch Alarm on canary metrics. Metrics are emitted by CloudWatch automatically and can be accessed by the following APIs:
- `canary.metricSuccessPercent()`
- `canary.metricDuration()`

> - `canary.metricFailed()`

Reviewer 4: Just calling out their names isn't helpful. The API reference doc has those, too. Include verbiage about what they mean and what people can use them for.

```ts

new cloudwatch.Alarm(this, 'CanaryAlarm', {
  metric: canary.metricSuccessPercent(),
  evaluationPeriods: 2,
  threshold: 90,
  comparisonOperator: cloudwatch.ComparisonOperator.LESS_THAN_THRESHOLD,
});
```

> ### Appendix: Canary CFN Property Implementation 

Reviewer 4: Why is this even there? What's the value I get from this as a user? The goal of the CDK is for me **not to have** to know any of this.

> ### Appendix: Canary CFN Property Implementation 

Reviewer 5: Again, this feels more like a design doc than a README for customers.

      var synthetics = require('Synthetics');
      const log = require('SyntheticsLogger');
  
      exports.handler = async function () {
        const requestOptions = {"hostname":"api.example.com","method":"","path":"/user/books/topbook/","port":443}
        let req = https.request(requestOptions);
        req.on('response', (res) => {
          log.info()
        });
      }`),
    handler: 'index.handler',
  }),
});
```

The canary will automatically produce a CloudWatch Dashboard:

![UI Screenshot](images/ui-screenshot.png)

### Canary Test Property 🚧

The `test` property represents the test that the canary executes. You can supply you own code using the `Test.custom()` method, or you can use a blueprint e.g `Test.heartbeat()`.

#### Custom 🚧

which will allow you to specify a custom script and handler for the canary. To specify the script in the `code` property, use the static method `code.fromInline()`.

```ts
const canary = new Canary(this, 'MyCanary', {
  test: Test.custom({ // 🚧 Not yet implemented 
    code: Code.fromInline('exports.handler = async () => {\nconsole.log(\'hello world\');\n};'),
    handler: 'index.handler',
  }),
});
```

> #### Blueprint 🚧

Reviewer 3: ### Blueprints 🚧

---
<!--END STABILITY BANNER-->

## Canaries

> AWS Cloudwatch Synthetics are used to create *canaries*, scripts that run on a schedule and monitor websites and API endpoints. You can read more about them [here](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries.html).

Reviewer 3: Amazon CloudWatch Synthetics allows you to monitor your application by continuously generating **synthetic** traffic, the traffic is generated by a **canary**, the simplest way to think of a canary is like a customer, a canary will hit your endpoint as if it was a customer allowing you to verify the health of your endpoint.

-   `s3:GetBucketLocation`
-   `s3:ListAllMyBuckets`
-   `cloudwatch:PutMetricData`
-   `logs:CreateLogGroup`
-   `logs:CreateLogStream`
-   `logs:PutLogEvents` 

`name` : The name of the canary. Must be lowercase and have no spaces. Redefined as `canaryName` in the Canary construct.

`runConfig` : The construct that contains imput about each canary run. The only property is `timeoutInSeconds`. The Canary construct replaces `runConfig` with the optional`timeout` property that defaults to the `rate` of the canary.

`runtimeVersion`: Specify the runtime version of the canary. The only valid runtime is `syn-1.0`. The Canary construct avoids future proofing and does not expose this property at all -- `syn-1.0` is hardcoded as the `runtimeVersion`. If more runtime versions are possible in the future, this can be changed to an enum-like property.

`schedule` : Specify how often the canary is to run and when the runs should stop. Type is a schedule object with two properties, `durationInSeconds` and `expression`. The Canary construct just exposes both as optional properties:
-  `lifetime` replaces `durationInSeconds` and defaults to `0` (canary runs forever until you stop it).

> - `rate` replaces `expression` and defaults to `rate(5 minutes)`. 

Author: what about changing `rate` to `frequency`? It seems like this property encapsulates how frequent the canary runs.

Author: this has been changed to `Schedule` to fit similar properties in autoscaling constructs.

});
```

#### From S3 Bucket

```ts
import * as synth from '@aws-cdk/aws-synthetics';
import * as s3 from '@aws-cdk/aws-s3';
import * as s3Deployment from '@aws-cdk/aws-s3-deployment';

// Create bucket
const bucket = new s3.Bucket(this, 'canarybucket');

// Upload assets from the canary directory to the bucket
new s3Deployment.BucketDeployment(this, 'Deployment', {
	sources: [s3Deployment.Source.asset('./canary')],
	destinationBucket: bucket,
});

// 'index.js' is the file name and key in s3
const canary = new synth.Canary(this,'my_test',{
	handler: 'index.handler',
	canaryName: 'mycanary',
	code: synth.Code.fromBucket(bucket,'index.js'),
});
```

### Canary Metrics

You can add metrics to your canary through the suite of `canary.metric()` APIs. They are: 

- `canary.metric(metricName, props?)`
- `canary.metricDuration(metricName, props?)`
- `canary.metricSucessPercent(metricName, props?)`
- `canary.metricFailed(metricName, props?)`

```ts
import * as synth from '@aws-cdk/aws-synthetics';

// 'index.js' is in the directory 'canary-handler'
const canary = new synth.Canary(this,'my_test',{
	handler: 'index.handler',
	canaryName: 'mycanary',

> 	code: synth.Code.fromAsset('./canary-handler'),

Reviewer 3: The use of `Code.fromAsset` before you explain it will confuse the reader, keep the `inline` from the pervious example

});

// define a metric on a canary

> canary.metricSuccessPercent();

Reviewer 3: We don't define a metric, this in itself is not interesting, the interesting part will be to show how you create an alarm on these metrics.
